### PR TITLE
Handle reset progress + pairing scenarios

### PR DESCRIPTION
**Pull request by jamescodeorg**
This change updates our pairing logic and UX to handle when a student in a pairing group deletes their progress.  Briefly, the behavior that we are targeting is as follows:
1) When the driver's progress is deleted, all navigators in the group retain their progress on the level and can tell that they were paired, but they can no longer tell who they were paired with and cannot see the source code that the group wrote.
2) When a navigator's progress is deleted, the driver and other navigators in the group retain their progress on the level and can tell that they were paired, but they can no longer tell who they were paired with.  (Source code is still accessible to the other students in the pairing group.)

This behavior is not necessarily ideal but is a reasonable compromise given how we currently store pairing data.  See the functional and tech specs for more details.

This translates to the following changes on the back-end:
- When progress is reset, delete rows from the user_levels tables but leave the rows in the paired_user_levels tables.  (implemented in #42235)
- Update PairedUserLevel and UserLevel to handle when the user_levels row referenced in paired_user_levels was deleted (user_level.rb, paired_user_level.rb).
- Update all three places where pairing data is retrieved and returned to the front-end (script_level.rb, levels_helper.rb, and api_controller.rb)

A good place to start reviewing would be the changes to UserLevel and PairedUserLevel.  The intent here is that UserLevel now has (more) methods to query pairing status and details about the pairing group.  Some of these calls are delegated to PairedUserLevel which should be viewed as being "private" to UserLevel.  Part of the reason for this separation is that I hope to replace PairedUserLevel at some point in the future with something a bit more flexible that can handle more scenarios.

The front-end is updated to handle the changes to the data from the back-end, most notably the fact that driver and navigator names can now be null to represent when the name is unknown.

## Links

- [functional spec](https://docs.google.com/document/d/1OslRELEPkzVDbqoQb8FbIOHURgZAy4YZ8WcnNieQzzw/edit#bookmark=id.92vcnuauw041)
- [tech spec](https://docs.google.com/document/d/1iiExH1ZHHGEJKeca3vQ8zpLZx3LPK5SW00qpN6GdgxA/edit#heading=h.qc7jxhm9j0ek)
- [jira ticket](https://codedotorg.atlassian.net/browse/LP-2001) (includes investigation notes)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/progress/ProgressBubble.jsx [hunk 1]
```diff
@@ -94,7 +94,7 @@ export default class ProgressBubble extends React.Component {
       level.isLocked,
       level.isUnplugged,
       level.bonus,
-      false,
+      level.paired,
```
Comment (jamescodeorg) [inline apps/src/templates/progress/ProgressBubble.jsx:97]: This fixes the issue where pairing icons weren't being shown for bubbles in the teacher panel.  Bubbles in the header and unit overview page are still not showing the pairing icon and will be fixed separated (LP-2003).

File: apps/src/code-studio/components/progress/teacherPanel/types.js [hunk 2]
```diff
@@ -19,13 +19,19 @@ const lessonExtra = PropTypes.shape({
   nextLevelPath: PropTypes.string
 });
 
+// This is the shape of the data returned by summarize_for_teacher_panel in
+// script_level.rb.  Note that this data is also directly passed to code that
+// expects an object with shape levelWithProgressType (defined in
+// progress/progressTypes.js) so it must be compatible with that shape.
```
Comment (jamescodeorg) [inline apps/src/code-studio/components/progress/teacherPanel/types.js:25]: This coupling worries me a bit but I didn't want to tackle it as part of this work (which has already taken much too much time).

File: apps/src/code-studio/components/progress/teacherPanel/SelectedStudentInfo.jsx [hunk 3]
```diff
@@ -110,12 +117,8 @@ export default class SelectedStudentInfo extends React.Component {
...
             <div>
               <div>{i18n.workedWith()}</div>
-              {navigator && (
-                <div key={navigator}>{i18n.partner({partner: navigator})}</div>
-              )}
-              {driver && (
-                <div key={driver}>{i18n.loggedIn({partner: driver})}</div>
-              )}
+              {isDriver && <div>{i18n.partner({partner: partner})}</div>}
+              {isNavigator && <div>{i18n.loggedIn({partner: partner})}</div>}
```
Comment (jamescodeorg) [inline apps/src/code-studio/components/progress/teacherPanel/SelectedStudentInfo.jsx:121]: There were supposed to be more changes to the UX here (e.g. to display the # of students in the pairing group) but the proposed strings didn't fit the space very well so I'm keeping the existing strings for now.

File: apps/src/StudioApp.js [hunk 4]
```diff
@@ -643,8 +643,15 @@ StudioApp.prototype.initProjectTemplateWorkspaceIconCallout = function() {
...
+// When pairing, source code is stored only with the driver. If the user completed
+// this level as a navigator, show an alert with a link to the (read-only) source
+// code stored in the driver's account.
 StudioApp.prototype.alertIfCompletedWhilePairing = function(config) {
-  if (!!config.level.pairingDriver) {
+  if (!config.level.isNavigator) {
+    return;
+  }
+
+  if (config.level.pairingDriver) {
```
Comment (maureensturgeon) [inline apps/src/StudioApp.js:654]: Is `pairingDriver` populated if the driver is logged in? If so, this alert would have displayed for the driver before as well, which it wouldn't now. Just want to make sure this is the intended behavior.

File: dashboard/app/models/paired_user_level.rb [hunk 5]
```diff
@@ -60,4 +60,38 @@ def self.pairs_by_user(users)
         memo
       end
   end
+
+  # Returns the User object representing the driver. Note that this returns nil
+  # if the driver's user account or the driver's progress was deleted.
+  def driver
```
Comment (maureensturgeon) [inline dashboard/app/models/paired_user_level.rb]: I would kind of expect this to be defined as an association like `belongs_to :driver, class_name: 'User', foreign_key: :user_id`

Comment (maureensturgeon) [reply dashboard/app/models/paired_user_level.rb]: Oops sorry, it would have to be through user_level.. let me try again. `has_one :driver, through: :driver_user_level, source: :user` works

File: dashboard/app/models/paired_user_level.rb [hunk 6]
```diff
@@ -60,4 +60,38 @@ def self.pairs_by_user(users)
...
+    # Get user_ids of navigator. Array may include nil elements if the
+    # navigator's progress was deleted.
+    navigator_user_ids = PairedUserLevel.
+      left_joins(:navigator_user_level).
+      where(driver_user_level_id: driver_user_level_id).
+      where.not(navigator_user_level_id: excluded_navigator_user_level_ids).
+      order('paired_user_levels.id DESC').
+      pluck('user_levels.user_id')
+
+    names_by_user_id = User.where(id: navigator_user_ids.compact).pluck(:id, :name).to_h
```
Comment (maureensturgeon) [inline dashboard/app/models/paired_user_level.rb]: I'm curious why you returned a hash to iterate over on the next line rather than returning the names directly from this query

File: dashboard/app/models/script_level.rb [hunk 7]
```diff
@@ -575,9 +572,11 @@ def summarize_for_teacher_panel(student, teacher = nil)
       id: level.id.to_s,
       contained: contained,
       submitLevel: level.properties['submittable'] == 'true',
-      paired: paired,
+      paired: is_driver || is_navigator || false,
```
Comment (maureensturgeon) [inline dashboard/app/models/script_level.rb:575]: Could this just be `is_driver || is_navigator`?

File: dashboard/app/helpers/levels_helper.rb [hunk 8]
```diff
@@ -248,14 +248,20 @@ def app_options
     end
 
     if pairing_check_user
-      recent_driver, recent_attempt, recent_user = UserLevel.most_recent_driver(@script, @level, pairing_check_user)
-      if recent_driver && !recent_user.is_a?(DeletedUser)
-        level_view_options(@level.id, pairing_driver: recent_driver)
-        if recent_attempt
-          level_view_options(@level.id, pairing_attempt: edit_level_source_path(recent_attempt)) if recent_attempt
+      user_level = UserLevel.find_by(user: pairing_check_user, script: @script, level: @level)
+      is_navigator = !user_level.nil? && user_level.navigator?
```
Comment (maureensturgeon) [inline dashboard/app/helpers/levels_helper.rb:252]: `!user_level.nil?` could be `user_level.present?`

File: dashboard/app/models/script_level.rb (hunk 7 above)
Comment (jamescodeorg) [reply dashboard/app/models/script_level.rb:575]: I wanted this to be a boolean (and not nil) when user_level is nil above.  I could change this to `!!(is_driver || is_navigator)` if that's clearer?

File: dashboard/app/controllers/api_controller.rb [hunk 9]
```diff
@@ -432,24 +432,29 @@ def user_progress_for_lesson
...
-          response[:pairingChannelId] = recent_channel if recent_channel
+        # Pairing info
+        is_navigator = user_level.navigator?
+        if is_navigator
+          driver = user_level.driver
+          driver_level_source_id = user_level.driver_level_source_id
+        end
+
+        response[:isNavigator] = is_navigator
+        if driver
```
Comment (maureensturgeon) [inline dashboard/app/controllers/api_controller.rb:451]: Is `driver` still set if the current user is the driver?

File: dashboard/app/models/paired_user_level.rb (hunk 6 above)
Comment (jamescodeorg) [reply dashboard/app/models/paired_user_level.rb]: Good question! The reason is because the user may have been deleted and I want to keep 'nil' placeholders in the returned array for those users.  This way, the count of entries in the returned array will always reflect the number of navigators in the group, even if users/progress have been deleted.

File: dashboard/app/models/user_level.rb [hunk 10]
```diff
@@ -67,42 +67,71 @@ def passing?
...
-  has_many :paired_user_levels_as_navigator, class_name: 'PairedUserLevel', foreign_key: 'navigator_user_level_id'
-  has_many :paired_user_levels_as_driver, class_name: 'PairedUserLevel', foreign_key: 'driver_user_level_id'
-
-  has_many :navigator_user_levels, through: :paired_user_levels_as_driver
-  has_many :driver_user_levels, through: :paired_user_levels_as_navigator
+  # Retrieves and memoizes the latest PairedUserLevel that's associated with
+  # this UserLevel for internal callers. External callers should call one of
+  # the higher-level pairing-related methods below.
+  #
+  # Conceptually, each UserLevel should only be associated with one
```
Comment (maureensturgeon) [inline dashboard/app/models/user_level.rb:74]: This explanation is really helpful!

File: dashboard/app/models/user_level.rb [hunk 11]
```diff
@@ -67,42 +67,71 @@ def passing?
...
+  # paired_user_levels table when posting progress, there can be multiple
+  # entries in the paired_user_levels table associated with a user_level. We
+  # heuristically consider the latest entry in the paired_user_levels table as
+  # the "active" one. This is correct for most cases but can be incorrect in
+  # some edge cases such as when a student leaves a pairing group and makes
+  # further progress on a level as an individual.
+  private def latest_paired_user_level
+    return @latest_paired_user_level if defined? @latest_paired_user_level
+    @latest_paired_user_level =
+      PairedUserLevel.where(driver_user_level_id: id).
```
Comment (maureensturgeon) [inline dashboard/app/models/user_level.rb:85]: Does `PairedUserLevel.where(driver_user_level_id: id).or(PairedUserLevel.where(navigator_user_level_id: id))` return them ordered by latest entry last?

File: dashboard/app/models/user_level.rb [hunk 12]
```diff
@@ -67,42 +67,71 @@ def passing?
...
+        or(PairedUserLevel.where(navigator_user_level_id: id)).
+        last
+  end
 
+  # Returns whether this UserLevel represents progress completed by a pairing
+  # group where the user was the driver.
   def driver?
-    navigator_user_levels.present?
+    !latest_paired_user_level.nil? &&
```
Comment (maureensturgeon) [inline dashboard/app/models/user_level.rb]: `!latest_paired_user_level.nil?` could be `latest_paired_user_level.present?`

File: dashboard/app/models/user_level.rb [hunk 13]
```diff
@@ -67,42 +67,71 @@ def passing?
...
+      id == latest_paired_user_level.driver_user_level_id
   end
 
+  # Returns whether this UserLevel represents progress completed by a pairing
+  # group where the user was a navigator.
   def navigator?
-    driver_user_levels.present?
+    !latest_paired_user_level.nil? &&
```
Comment (maureensturgeon) [inline dashboard/app/models/user_level.rb]: Same comment as above

File: apps/src/StudioApp.js (hunk 4 above)
Comment (jamescodeorg) [reply apps/src/StudioApp.js:654]: No, the previous implementation would return nil for pairingDriver.  It went through UserLevel#driver_user_levels which would be nil for the driver's user_level.

One thing to note is, previously, existence of pairingDriver indicated whether the the current user was a navigator.  I had to introduce isNavigator to distinguish the case where the current user is a navigator but we don't have the driver's name.

File: dashboard/app/controllers/api_controller.rb (hunk 9 above)
Comment (jamescodeorg) [reply dashboard/app/controllers/api_controller.rb:451]: No, because driver is set above on in line 446 inside the `is_navigator`.  I'm open to other ways of structuring this if this is confusing?

File: dashboard/app/helpers/levels_helper.rb (hunk 8 above)
Comment (jamescodeorg) [reply dashboard/app/helpers/levels_helper.rb:252]: I didn't use `.present?` because it's not equivalent to `! .nil?` for empty string, empty object, empty array and some other cases.  I'm really surprised that `.not_nil?` was never added to ruby or rails -- I feel like this is a very basic operation that I want to do all the time!

File: dashboard/app/models/user_level.rb (hunk 11 above)
Comment (jamescodeorg) [reply dashboard/app/models/user_level.rb:85]: This query returns values in the default order which is by id.  This is effectively the same as ordering by creation/modification time (ascending) because entries in this table are never modified.

File: dashboard/app/models/user_level.rb (hunk 12 above)
Comment (jamescodeorg) [reply dashboard/app/models/user_level.rb]: (see comment above about `! .nil?` vs `.present?`.  Let's discuss if you disagree!)

File: apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx [hunk 14]
```diff
@@ -72,27 +74,63 @@ describe('SelectedStudentInfo', () => {
...
+    expect(wrapper.contains('Last Updated:')).to.equal(true);
+  });
+
+  it('displays time and unknown navigator if paired as driver on level', () => {
+    const levelWithProgress = {
+      ...LEVEL_WITH_PROGRESS,
+      paired: true,
+      isDriver: true,
+      status: LevelStatus.perfect,
+      navigators: [null, 'Student 3']
```
Comment (maureensturgeon) [inline apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx]: Now that I'm seeing this, it would be nice if the backend passed back the placeholder name. But we'd have to make that placeholder like `n/a` translatable.. It seems strange to pass an array around where null actually means something. Have you considered other options for this?

File: apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx [hunk 15]
```diff
@@ -72,27 +74,63 @@ describe('SelectedStudentInfo', () => {
...
+    const levelWithProgress = {
+      ...LEVEL_WITH_PROGRESS,
+      paired: true,
+      isNavigator: true,
+      status: LevelStatus.perfect,
+      driver: null
+    };
+    const wrapper = setUp({levelWithProgress});
+
+    expect(wrapper.contains('Worked With:')).to.equal(true);
```
Comment (maureensturgeon) [inline apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx:131]: Nit: these could be `.to.be.true` and `.to.be.false` instead of using `.equal`

File: dashboard/test/models/user_level_test.rb [hunk 16]
```diff
@@ -51,6 +54,36 @@ class UserLevelTest < ActiveSupport::TestCase
...
+  end
+
+  test "driver" do
+    assert_equal nil, @unpaired_user_level.driver
+    assert_equal @driver, @driver_user_level.driver
+    assert_equal @driver, @navigator_user_level.driver
+  end
+
+  test "navigators_names" do
+    assert_equal nil, @unpaired_user_level.navigators_names
```
Comment (maureensturgeon) [inline dashboard/test/models/user_level_test.rb:82]: Would it be better if this method just returned an empty array instead of nil? So that the type of the thing it returns is predictable?

**Review by maureensturgeon — COMMENTED** [review]
Looks good! I just have lots of questions :)

File: dashboard/app/models/script_level.rb (hunk 7 above)
Comment (maureensturgeon) [reply dashboard/app/models/script_level.rb:575]: Oh, got it! No I think this is fine, I just didn't think of that. It would be great to cover this with a unit test to make sure someone doesn't accidentally refactor it to `is_driver || is_navigator` (like me in the future)

File: apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx (hunk 14 above)
Comment (jamescodeorg) [reply apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx]: Yeah, this is something I went back and forth on a couple of times.  I thought that it would be better for the back-end to return an indication that the data was missing and let the front-end decide how to display that.  I couldn't think of a better way to express "don't have this value" in JSON.  One possibility would be to pass a separate array of boolean flags (or similarly, return an array of objects which include the flag and the name) but that seemed too cumbersome?  Any ideas?

(In general, I'm not super comfortable with how much meaning I'm assigning to `nil` in this change.  I think using an [Option/Maybe pattern](https://medium.com/bettersharing/option-pattern-in-ruby-7b0f7c5abdb6) would be more robust but it feels kind of clunky in Ruby and I didn't want to introduce it as part of this change.)

File: dashboard/app/helpers/levels_helper.rb (hunk 8 above)
Comment (maureensturgeon) [reply dashboard/app/helpers/levels_helper.rb:252]: The way it's implemented, user_level would only be an active record or nil, right? So wouldn't it be safe to use in this case?

File: apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx (hunk 15 above)
Comment (jamescodeorg) [reply apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx:131]: Yeah, agreed.  I was trying to match the style of the code that was already there.  Do you think I should go ahead and clean up the whole file?

File: dashboard/test/models/user_level_test.rb (hunk 16 above)
Comment (jamescodeorg) [reply dashboard/test/models/user_level_test.rb:82]: Good question.  I chose nil here because I was trying to keep it consistent with `driver` which returns nil when the user_level is unpaired.  Thoughts?

File: dashboard/app/controllers/api_controller.rb (hunk 9 above)
Comment (maureensturgeon) [reply dashboard/app/controllers/api_controller.rb:451]: If the current user is the driver, then we don't set `pairingDriver`?

File: dashboard/app/helpers/levels_helper.rb (hunk 8 above)
Comment (jamescodeorg) [reply dashboard/app/helpers/levels_helper.rb:252]: Yes, I agree that it's safe for the current implementation but there's no way to guarantee the type of user_level in the future and (to me) `! .nil?` expresses the intent better than `.present?`.  Could you elaborate a bit on why you prefer `.present?`?

File: apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx (hunk 14 above)
Comment (maureensturgeon) [reply apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx]: You could pass `annonymousNavigatorCount`? It's more data to pass around, but it more clearly communicates intent. I think having an array of nulls could easily confuse someone

File: apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx (hunk 15 above)
Comment (maureensturgeon) [reply apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx:131]: Up to you! I don't have a strong opinion

File: dashboard/app/controllers/api_controller.rb (hunk 9 above)
Comment (jamescodeorg) [reply dashboard/app/controllers/api_controller.rb:451]: If the current user is the driver, we do not want to set `pairingDriver` in the response to the client. 

This used to be more important because the client used the existence of `pairingDriver` to determine if the current user was the navigator.  There is now an explicit flag that indicates if the current user is a navigator so it would be ok to always send `pairingDriver` but I left the previous behavior that `pairingDriver`, `pairingAttempt`, and `pairingChannelId` are only sent if the current user is the navigator.

File: dashboard/app/helpers/levels_helper.rb (hunk 8 above)
Comment (maureensturgeon) [reply dashboard/app/helpers/levels_helper.rb:252]: `.present?` is easier to understand than `! .nil?`, I try to avoid double negatives in general because they're harder to reason about

File: dashboard/test/models/user_level_test.rb (hunk 16 above)
Comment (maureensturgeon) [reply dashboard/test/models/user_level_test.rb:82]: I slightly learn towards a preference to return an empty array for consistency, but this is really a nit pick. Feel free to leave as-is

File: dashboard/app/helpers/levels_helper.rb (hunk 8 above)
Comment (jamescodeorg) [reply dashboard/app/helpers/levels_helper.rb:252]: I started a Slack thread about this: https://codedotorg.slack.com/archives/C0T0PNTM3/p1632422797381300.  Let's try Maddie's suggestion?

File: dashboard/test/models/user_level_test.rb (hunk 16 above)
Comment (jamescodeorg) [reply dashboard/test/models/user_level_test.rb:82]: I think I favor the consistency with `driver` more so I'll leave this as-is.

File: dashboard/app/helpers/levels_helper.rb (hunk 8 above)
Comment (jamescodeorg) [reply dashboard/app/helpers/levels_helper.rb:252]: Maureen and I talked through a bunch of options offline and decided that this is ok to keep as-is.

File: apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx (hunk 14 above)
Comment (jamescodeorg) [reply apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx]: I ended up creating a separate method for (total) navigator count.  I like this much better than what I had originally, thanks for pushing on this!  Note that the count isn't fully plumbed through to the front-end yet.  I'll do that in a future PR after I discuss the UX with Amanda.

Comment (maureensturgeon) [reply apps/test/unit/code-studio/components/progress/teacherPanel/SelectedStudentInfoTest.jsx]: Great solution!

**Review by maureensturgeon — APPROVED** [review]
Thanks for walking through all the questions with me, I appreciate it!!
